### PR TITLE
Core arbiter hyper twin aware

### PR DESCRIPTION
**Pull request by kraftp**
I've also addressed all of your comments on NO_ARBITER_in_runtime.  I'll probably end up merging NO_ARBITER_in_runtime into master first, then merge this into master next in two separate commits for clarity.  The other part of this PR is in PerfUtils.

### CODE REVIEW

File: src/CoreArbiterServer.h [hunk 1]
```diff
@@ -82,6 +83,10 @@ class CoreArbiterServer {
         // thread on this core that ran sched_getcpu().
         int id;
 
+        // The ID of this core's hypertwin. This ID matches what would be returned by a
+        // thread on the hypertwin that ran sched_getcpu().
+        int hypertwin_id;
```
Comment (johnousterhout) [inline src/CoreArbiterServer.h]: No underscores in variable names: camelCase.

File: src/CoreArbiterServer.cc [hunk 2]
```diff
@@ -1017,11 +1016,28 @@ CoreArbiterServer::distributeCores()
...
-        unmanagedCores.erase(unmanagedCores.end() - numCoresToMakeManaged,
-                              unmanagedCores.end());
+        // Ensure hypertwin of managed core not managed unless unavoidable.
+        for (int i = 0; i < (int) numCoresToMakeManaged; i++) {
+
+            int coreIdToManageIndex = 0;
+            for (int j = 0; j < (int) unmanagedCores.size(); j++) {
+                bool match = false;
+                for (int k = 0; k < (int) managedCores.size(); 
+                    k++) {
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: Continuation lines should have double the normal indent.

File: src/CoreArbiterServer.cc [hunk 3]
```diff
@@ -1017,11 +1016,28 @@ CoreArbiterServer::distributeCores()
...
-                              unmanagedCores.end() - numCoresToMakeManaged,
-                              unmanagedCores.end());
-        unmanagedCores.erase(unmanagedCores.end() - numCoresToMakeManaged,
-                              unmanagedCores.end());
+        // Ensure hypertwin of managed core not managed unless unavoidable.
+        for (int i = 0; i < (int) numCoresToMakeManaged; i++) {
+
+            int coreIdToManageIndex = 0;
+            for (int j = 0; j < (int) unmanagedCores.size(); j++) {
+                bool match = false;
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc]: "Match" is not a great variable name; I can't tell from the code what it means, other than "random boolean variable".

File: src/CoreArbiterServer.cc [hunk 4]
```diff
@@ -1017,11 +1016,28 @@ CoreArbiterServer::distributeCores()
         size_t numCoresToMakeManaged =
             numAssignedCores - managedCores.size();
         LOG(NOTICE, "Making %lu cores managed", numCoresToMakeManaged);
-        managedCores.insert(managedCores.end(),
-                              unmanagedCores.end() - numCoresToMakeManaged,
-                              unmanagedCores.end());
-        unmanagedCores.erase(unmanagedCores.end() - numCoresToMakeManaged,
-                              unmanagedCores.end());
+        // Ensure hypertwin of managed core not managed unless unavoidable.
```
Comment (johnousterhout) [inline src/CoreArbiterServer.cc:1019]: I don't believe this is the behavior we want. What we want is for the hypertwin to be managed but idle. If you leave it unmanaged, the kernel will run other stuff there, which will interfere with the thread on the other hypertwin. Let's discuss.

File: src/CoreArbiterServer.h [hunk 5]
```diff
@@ -103,6 +108,7 @@ class CoreArbiterServer {
 
         CoreInfo(int id, std::string managedTasksPath)
             : id(id)
+            , hypertwin_id(PerfUtils::Util::getHyperTwin(id))
```
Comment (johnousterhout) [inline src/CoreArbiterServer.h]: Is there a pull request for the PerfUtils package (or has getHyperTwin always been present there)?

**Comment by hq6** [comment]
Most of this pull has been merged in; discard the rest.

**Pull request closed without merging**